io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -1079,130 +1090,290 @@ traceMany ((time, tid, event):ts) trace =
-      --TODO: avoid duplicates!
-      modifySTRef blocked (mytid:)
-
-
-execAtomically' :: StmA s () -> ST s [(ThreadId, [TVarId])]
-execAtomically' = go []
+    go :: forall b.
+          StmStack s b a
+       -> Map TVarId (SomeTVar s)  -- set of vars read
+       -> Map TVarId (SomeTVar s)  -- set of vars written
+       -> [SomeTVar s]             -- vars written in order (no dups)
+       -> TVarId                   -- var fresh name supply
+       -> StmA s b
+       -> ST s (StmTxResult s a)
+    go ctl !read !written writtenSeq !nextVid action = assert localInvariant $
+                                                       case action of
+      ReturnStm x -> case ctl of
+        AtomicallyFrame -> do
+          -- Commit each TVar
+          traverse_ (\(SomeTVar tvar) -> do
+                        commitTVar tvar
+                        -- Also assert the data invariant that outside a tx
+                        -- the undo stack is empty:
+                        undos <- readTVarUndos tvar
+                        assert (null undos) $ return ()
+                    ) written
+
+          -- Return the vars written, so readers can be unblocked
+          return (StmTxCommitted x (reverse writtenSeq) nextVid)
+
+        OrElseLeftFrame _b k writtenOuter writtenOuterSeq ctl' -> do
+          -- Commit the TVars written in this sub-transaction that are also
+          -- in the written set of the outer transaction
+          traverse_ (\(SomeTVar tvar) -> commitTVar tvar)
+                    (Map.intersection written writtenOuter)
+          -- Merge the written set of the inner with the outer
+          let written'    = Map.union written writtenOuter
+              writtenSeq' = filter (\(SomeTVar tvar) ->
+                                      tvarId tvar `Map.notMember` writtenOuter)
+                                    writtenSeq
+                         ++ writtenOuterSeq
+          -- Skip the orElse right hand and continue with the k continuation
+          go ctl' read written' writtenSeq' nextVid (k x)
+
+        OrElseRightFrame k writtenOuter writtenOuterSeq ctl' -> do
+          -- Commit the TVars written in this sub-transaction that are also
+          -- in the written set of the outer transaction
+          traverse_ (\(SomeTVar tvar) -> commitTVar tvar)
+                    (Map.intersection written writtenOuter)
+          -- Merge the written set of the inner with the outer
+          let written'    = Map.union written writtenOuter
+              writtenSeq' = filter (\(SomeTVar tvar) ->
+                                      tvarId tvar `Map.notMember` writtenOuter)
+                                    writtenSeq
+                         ++ writtenOuterSeq
+          -- Continue with the k continuation
+          go ctl' read written' writtenSeq' nextVid (k x)
+
+      ThrowStm e -> do
+        -- Revert all the TVar writes
+        traverse_ (\(SomeTVar tvar) -> revertTVar tvar) written
```

Comment:
I don't this this really matters, but we violate `OR2` if we revert all the written `TVar`s.  If I comment this out, the tests still pass.

Reply:
Finding violations is a bit tricky, my generators are not perfect. I have had to run with 10,000 tests to find failures, but I've also checked with 100,000 passes.